# WE04 Universal Bank

## Assignment Submitted by Aravind Dudam

## U04440266

## Importing libraries

In [42]:
import pandas as pd
import numpy as np
from sklearn.svm import SVC
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
# set random seed to ensure that results are repeatable
np.random.seed(1)

## Loading dataset

In [43]:
# Loading the data
data = pd.read_csv("E:/DSP/Week 04/UniversalBank.csv")

In [44]:
# Show the data
data.head

<bound method NDFrame.head of         ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  \
0        1   25           1      49     91107       4    1.6          1   
1        2   45          19      34     90089       3    1.5          1   
2        3   39          15      11     94720       1    1.0          1   
3        4   35           9     100     94112       1    2.7          2   
4        5   35           8      45     91330       4    1.0          2   
...    ...  ...         ...     ...       ...     ...    ...        ...   
4995  4996   29           3      40     92697       1    1.9          3   
4996  4997   30           4      15     92037       4    0.4          1   
4997  4998   63          39      24     93023       2    0.3          3   
4998  4999   65          40      49     90034       3    0.5          2   
4999  5000   28           4      83     92612       3    0.8          1   

      Mortgage  Personal Loan  Securities Account  CD Account  Online

In [45]:
# Number of rows and columns
data.shape 

(5000, 14)

In [46]:
# Target Variable "CD Account"
data['CD Account'].value_counts()

0    4698
1     302
Name: CD Account, dtype: int64

In [47]:
# Identifying Nulls and Non Nulls
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  5000 non-null   int64  
 1   Age                 5000 non-null   int64  
 2   Experience          5000 non-null   int64  
 3   Income              5000 non-null   int64  
 4   ZIP Code            5000 non-null   int64  
 5   Family              5000 non-null   int64  
 6   CCAvg               5000 non-null   float64
 7   Education           5000 non-null   int64  
 8   Mortgage            5000 non-null   int64  
 9   Personal Loan       5000 non-null   int64  
 10  Securities Account  5000 non-null   int64  
 11  CD Account          5000 non-null   int64  
 12  Online              5000 non-null   int64  
 13  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 547.0 KB


## Check the missing values with na's for each variable

In [48]:
data.isna().sum()

ID                    0
Age                   0
Experience            0
Income                0
ZIP Code              0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal Loan         0
Securities Account    0
CD Account            0
Online                0
CreditCard            0
dtype: int64

In [49]:
data.isnull().sum()

ID                    0
Age                   0
Experience            0
Income                0
ZIP Code              0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal Loan         0
Securities Account    0
CD Account            0
Online                0
CreditCard            0
dtype: int64

In [50]:
data.corr()['CD Account'].sort_values(ascending=False)

CD Account            1.000000
Securities Account    0.317034
Personal Loan         0.316355
CreditCard            0.278644
Online                0.175880
Income                0.169738
CCAvg                 0.136534
Mortgage              0.089311
ZIP Code              0.019972
Family                0.014110
Education             0.013934
Experience            0.010353
Age                   0.008043
ID                   -0.006909
Name: CD Account, dtype: float64

## Data Cleaning

In [51]:
data.drop(columns=["ID"])
data.drop(columns=["ZIP Code"])

ID  Age  Experience  Income  Family  CCAvg  Education  Mortgage  \
0        1   25           1      49       4    1.6          1         0   
1        2   45          19      34       3    1.5          1         0   
2        3   39          15      11       1    1.0          1         0   
3        4   35           9     100       1    2.7          2         0   
4        5   35           8      45       4    1.0          2         0   
...    ...  ...         ...     ...     ...    ...        ...       ...   
4995  4996   29           3      40       1    1.9          3         0   
4996  4997   30           4      15       4    0.4          1        85   
4997  4998   63          39      24       2    0.3          3         0   
4998  4999   65          40      49       3    0.5          2         0   
4999  5000   28           4      83       3    0.8          1         0   

      Personal Loan  Securities Account  CD Account  Online  CreditCard  
0                 0                   1           0       0           0  
1                 0                   1           0       0           0  
2                 0                   0           0       0           0  
3                 0                   0           0       0           0  
4                 0                   0           0       0           1  
...             ...                 ...         ...     ...         ...  
4995              0                   0           0       1           0  
4996              0                   0           0       1           0  
4997              0                   0           0       0           0  
4998              0                   0           0       1           0  
4999              0                   0           0       1           1  

[5000 rows x 13 columns]

In [52]:
# The dataset is cleaned with the necessary columns.

In [53]:
# Performing One hot encoding for preprocessing Education column

## Using One-hot Encoding to encode the categorical variable "Education"

In [54]:
EncodedData = pd.get_dummies(data, columns = ['Education'])
EncodedData

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Mortgage  \
0        1   25           1      49     91107       4    1.6         0   
1        2   45          19      34     90089       3    1.5         0   
2        3   39          15      11     94720       1    1.0         0   
3        4   35           9     100     94112       1    2.7         0   
4        5   35           8      45     91330       4    1.0         0   
...    ...  ...         ...     ...       ...     ...    ...       ...   
4995  4996   29           3      40     92697       1    1.9         0   
4996  4997   30           4      15     92037       4    0.4        85   
4997  4998   63          39      24     93023       2    0.3         0   
4998  4999   65          40      49     90034       3    0.5         0   
4999  5000   28           4      83     92612       3    0.8         0   

      Personal Loan  Securities Account  CD Account  Online  CreditCard  \
0                 0                   1           0       0           0   
1                 0                   1           0       0           0   
2                 0                   0           0       0           0   
3                 0                   0           0       0           0   
4                 0                   0           0       0           1   
...             ...                 ...         ...     ...         ...   
4995              0                   0           0       1           0   
4996              0                   0           0       1           0   
4997              0                   0           0       0           0   
4998              0                   0           0       1           0   
4999              0                   0           0       1           1   

      Education_1  Education_2  Education_3  
0               1            0            0  
1               1            0            0  
2               1            0            0  
3               0            1            0  
4               0            1            0  
...           ...          ...          ...  
4995            0            0            1  
4996            1            0            0  
4997            0            0            1  
4998            0            1            0  
4999            1            0            0  

[5000 rows x 16 columns]

In [55]:
# Data split - train and test

In [56]:
train_df, test_df = train_test_split(data, test_size=0.3)
target = 'CD Account'
predictors = list(data.columns)
predictors.remove(target)

In [57]:
scaler = preprocessing.StandardScaler()
cols_to_stdize = predictors               
train_df[cols_to_stdize] = scaler.fit_transform(train_df[cols_to_stdize])
test_df[cols_to_stdize] = scaler.transform(test_df[cols_to_stdize])

train_X = train_df[predictors]
train_y = train_df[target]
test_X = train_df[predictors]
test_y = test_df[target]

In [65]:
scaler = MinMaxScaler()
model=scaler.fit(EncodedData)
scaled_data=model.transform(EncodedData)

scaled_data = pd.DataFrame(scaled_data, index=EncodedData.index, columns=EncodedData.columns)

In [66]:
X = scaled_data.drop('CD Account',axis=1)
y = scaled_data['CD Account']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 14)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((3500, 15), (1500, 15), (3500,), (1500,))

# Modeling the data

## Linear regression with Randomsearch

In [60]:
score_measure = "recall"
kfolds = 5

param_grid = {'C':[0.001,0.01,0.1,1,10], # C is the regulization strength
               'penalty':['l1', 'l2','elasticnet','none'],
              'solver':['saga','liblinear'],
              'max_iter': np.arange(500,1000)
    
    
}

lr = LogisticRegression()
rand_search = RandomizedSearchCV(estimator = lr, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = rand_search.fit(train_X,train_y)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestlr = rand_search.best_estimator_

Fitting 5 folds for each of 500 candidates, totalling 2500 fits
The best recall score is 0.6708245243128964
... with parameters: {'solver': 'liblinear', 'penalty': 'l2', 'max_iter': 890, 'C': 0.1}


C:\Users\aravi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
925 fits failed out of a total of 2500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
330 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\aravi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\aravi\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py", line 1461, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
  File "C:\Users\aravi\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py", line 464, in _check_solver
    raise ValueError("

## Logistic regression using grid search

In [61]:
score_measure = "recall"
kfolds = 5
best_penality = rand_search.best_params_['penalty']
best_solver = rand_search.best_params_['solver']
min_regulization_strength=rand_search.best_params_['C']
min_iter = rand_search.best_params_['max_iter']

#Using the best parameters from the Random Search to use as range for the parameters to do the grid search
param_grid = {
    
    'C':np.arange(min_regulization_strength-1,min_regulization_strength+1), 
               'penalty':[best_penality],
              'solver':[best_solver],
              'max_iter': np.arange(min_iter-400,min_iter+400)
}

logreg =  LogisticRegression()
grid_search = GridSearchCV(estimator = logreg, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1, # n_jobs=-1 will utilize all available CPUs 
                return_train_score=True)

_ = grid_search.fit(train_X,train_y)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestlogreg = grid_search.best_estimator_

Fitting 5 folds for each of 1600 candidates, totalling 8000 fits
The best recall score is 0.6708245243128964
... with parameters: {'C': 0.09999999999999998, 'max_iter': 490, 'penalty': 'l2', 'solver': 'liblinear'}


C:\Users\aravi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
4000 fits failed out of a total of 8000.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
4000 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\aravi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\aravi\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py", line 1464, in fit
    raise ValueError("Penalty term must be positive; got (C=%r)" % self.C)
ValueError: Penalty term must be positive; got (C=-0.9)

  warnings.warn(some_fits_failed_message, FitFailedWarning)
C:\User

## SVM classification model using Random Search for Linear kernel

In [70]:
score_measure = "recall"
kfolds = 5

param_grid = {
     'C': [0.0001, 0.001, 0.1, 1, 10, 100, 1000], 
    'kernel': ['linear']
}

random_linear_SVC = SVC()
random_search = RandomizedSearchCV(estimator = random_linear_SVC, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = random_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {random_search.best_score_}")
print(f"... with parameters: {random_search.best_params_}")

bestPrecisionTree = random_search.best_estimator_

C:\Users\aravi\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 7 is smaller than n_iter=500. Running 7 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 5 folds for each of 7 candidates, totalling 35 fits
The best recall score is 0.6691056910569106
... with parameters: {'kernel': 'linear', 'C': 0.1}


## SVM classification model using Grid Search for Linear kernel

In [71]:
score_measure = "recall"
kfolds = 5

param_grid = {
     'C': [0.0001, 0.001, 0.1, 1, 10, 100, 1000], 
    'kernel': ['linear']
}

grid_linear_SVC = SVC()
grid_search = GridSearchCV(estimator = grid_linear_SVC, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestPrecisionTree = grid_search.best_estimator_

Fitting 5 folds for each of 7 candidates, totalling 35 fits
The best recall score is 0.6691056910569106
... with parameters: {'C': 0.1, 'kernel': 'linear'}


## SVM classification model using Random Search for Poly kernel

In [73]:
score_measure = "recall"
kfolds = 5

param_grid = {
     'C': [0.1,1, 10, 100], 
    'gamma': [1,0.1,0.01,0.001],
    'kernel': ['poly']
}

random_Poly_SVC = SVC()
random_search = RandomizedSearchCV(estimator = random_Poly_SVC, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = random_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {random_search.best_score_}")
print(f"... with parameters: {random_search.best_params_}")

bestPrecisionTree = rand_search.best_estimator_

C:\Users\aravi\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 16 is smaller than n_iter=500. Running 16 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 5 folds for each of 16 candidates, totalling 80 fits
The best recall score is 0.6980255516840883
... with parameters: {'kernel': 'poly', 'gamma': 1, 'C': 10}


## SVM classification model using Grid Search for Poly kernel

In [74]:
score_measure = "recall"
kfolds = 5

param_grid = {
     'C': [0.1,1, 10, 100], 
    'gamma': [1,0.1,0.01,0.001],
    'kernel': ['poly']
}


grid_Poly_SVC = SVC()
grid_search = GridSearchCV(estimator = grid_Poly_SVC, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestPrecisionTree = grid_search.best_estimator_

Fitting 5 folds for each of 16 candidates, totalling 80 fits
The best recall score is 0.6980255516840883
... with parameters: {'C': 10, 'gamma': 1, 'kernel': 'poly'}


## SVM classification model using Random Search for rbf kernel

In [76]:
score_measure = "recall"
kfolds = 10

param_grid = {
     'C': [0.1,1, 10], 
    'gamma': [1,0.1,0.011],
    'kernel': ['rbf']
}

random_rbf_SVC = SVC()
random_search = RandomizedSearchCV(estimator = random_rbf_SVC, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = random_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {random_search.best_score_}")
print(f"... with parameters: {random_search.best_params_}")

bestPrecisionTree = rand_search.best_estimator_

C:\Users\aravi\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 9 is smaller than n_iter=500. Running 9 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 10 folds for each of 9 candidates, totalling 90 fits
The best recall score is 0.6688095238095239
... with parameters: {'kernel': 'rbf', 'gamma': 0.1, 'C': 1}


## SVM classification model using Grid Search for rbf kernel

In [77]:
score_measure = "recall"
kfolds = 5

param_grid = {
     'C': [0.1,1, 10], 
    'gamma': [1,0.1,0.011],
    'kernel': ['rbf']
}

grid_rbf_SVC = SVC()
grid_search = GridSearchCV(estimator = grid_rbf_SVC, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestPrecisionTree = grid_search.best_estimator_

Fitting 5 folds for each of 9 candidates, totalling 45 fits
The best recall score is 0.6691056910569106
... with parameters: {'C': 1, 'gamma': 0.1, 'kernel': 'rbf'}


## Random search for the Decision Tree Classifier

In [78]:
score_measure = "recall"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(1,100),  
    'min_samples_leaf': np.arange(1,100),
    'min_impurity_decrease': np.arange(0.0001, 0.01, 0.0005),
    'max_leaf_nodes': np.arange(5, 100), 
    'max_depth': np.arange(1,50), 
    'criterion': ['entropy', 'gini'],
}

dtree = DecisionTreeClassifier()
rand_search = RandomizedSearchCV(estimator = dtree, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestRecallTree = rand_search.best_estimator_

Fitting 5 folds for each of 500 candidates, totalling 2500 fits
The best recall score is 0.7170731707317073
... with parameters: {'min_samples_split': 99, 'min_samples_leaf': 4, 'min_impurity_decrease': 0.0036, 'max_leaf_nodes': 93, 'max_depth': 7, 'criterion': 'entropy'}


C:\Users\aravi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
15 fits failed out of a total of 2500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
15 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\aravi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\aravi\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 937, in fit
    super().fit(
  File "C:\Users\aravi\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 250, in fit
    raise ValueError(
ValueError: min_samples_split must be an integer greater than 1 or a float in (

## Grid search for the Decision Tree Classifier

In [79]:
score_measure = "recall"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(94,99),  
    'min_samples_leaf': np.arange(9,13),
    'min_impurity_decrease': np.arange(0.0038, 0.0044, 0.0001),
    'max_leaf_nodes': np.arange(92,98), 
    'max_depth': np.arange(38,44), 
    'criterion': ['entropy'],
}

dtree = DecisionTreeClassifier()
grid_search = GridSearchCV(estimator = dtree, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestRecallTree = grid_search.best_estimator_

Fitting 5 folds for each of 5040 candidates, totalling 25200 fits
The best recall score is 0.7170731707317073
... with parameters: {'criterion': 'entropy', 'max_depth': 38, 'max_leaf_nodes': 92, 'min_impurity_decrease': 0.0038, 'min_samples_leaf': 9, 'min_samples_split': 94}


## Inference

From the above results, here are the best recall scores determined:
    
    1. Logistic Regression with the Random-Search search 67.1%
    2. Logistic Regression with the Grid search 67.1%
    3. SVM classification model using Random Search for Linear kernel is 66.9%
    4. SVM classification model using Grid Search for Linear kernel is 66.9%
    5. SVM classification model using Random Search for Poly kernel is 69.8%
    6. SVM classification model using Grid Search for Poly kernel is 69.8%
    7. SVM classification model using Random Search for rbf kernel is 66.8%
    8. SVM classification model using Grid Search for rbf kernel is 66.9%
    9. Random search for the Decision Tree Classifier is 71.7%
    10. Grid search for the Decision Tree Classifier is 71.7%
    
We have all of the values for the decision tree classifier with the optimal parameters for each model examined, including those for grid search, random search, and SVM Classification model employing three kernels (Linera, Poly, and rbf). This allows us to conclude that the decision tree classifier, for both randomSearch CV and GridSearch CV, has the best recall score, which is 71.7%. We may conclude that the decision tree classifier is the best model.